### PR TITLE
[keyboard]Graystudio Think6.5 V3 Support

### PR DESCRIPTION
**Pull request by edwardslau**
## Description

Think6.5V3 PCB Firmware
## Types of Changes

- [x] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/gray_studio/think65v3/config.h [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+#define RGBLIGHT_EFFECT_RAINBOW_SWIRL
+#define RGBLIGHT_EFFECT_SNAKE
+#define RGBLIGHT_EFFECT_KNIGHT
+#define RGBLIGHT_EFFECT_CHRISTMAS
+#define RGBLIGHT_EFFECT_STATIC_GRADIENT
+#define RGBLIGHT_EFFECT_RGB_TEST
+#define RGBLIGHT_EFFECT_ALTERNATING
+#define RGBLIGHT_EFFECT_TWINKLE
+#define RGBLIGHT_DEFAULT_MODE RGBLIGHT_MODE_RAINBOW_MOOD
+#endif
```
Comment (adophoxia) [inline keyboards/gray_studio/think65v3/config.h:44]:
```suggestion

```
Should be moved to `info.json`.

File: keyboards/gray_studio/think65v3/rules.mk [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+#
+BOOTMAGIC_ENABLE = yes     # Virtual DIP switch configuration
+MOUSEKEY_ENABLE = yes       # Mouse keys
+EXTRAKEY_ENABLE = yes       # Audio control and System control
+CONSOLE_ENABLE = no        # Console for debug
+COMMAND_ENABLE = no        # Commands for debug and configuration
+NKRO_ENABLE = yes            # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality on B7 by default
+RGBLIGHT_ENABLE = yes        # Enable keyboard RGB underglow
+AUDIO_ENABLE = no           # Audio output on port C6
```
Comment (adophoxia) [inline keyboards/gray_studio/think65v3/rules.mk:15]:
```suggestion

```
Should be moved to `info.json`.

File: keyboards/gray_studio/think65v3/think65v3.h [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    K20,    K22, K23, K24, K25, K26, K27, K28, K29, K2A, K2B, K2C, K1E, K2D,          \
+    K30,  K31, K32, K33, K34, K35, K36, K37, K38, K39, K3A, K3B, K3D,     K3E,        \
+    K40,  K41,  K42,                    K47,           K4A,  K4B,     K4D, K4E, K4F    \
+) { \
+    { K00,   K01,   K02,   K03,   K04,   K05,   K06,   K07,   K08,   K09,   K0A,   K0B,   K0C,   K0D,   KC_NO, K0F   }, \
+    { K10,   KC_NO, K12,   K13,   K14,   K15,   K16,   K17,   K18,   K19,   K1A,   K1B,   K1C,   K1D,   K1E,   K1F   }, \
+    { K20,   KC_NO, K22,   K23,   K24,   K25,   K26,   K27,   K28,   K29,   K2A,   K2B,   K2C,   K2D,   KC_NO, KC_NO }, \
+    { K30,   K31,   K32,   K33,   K34,   K35,   K36,   K37,   K38,   K39,   K3A,   K3B,   KC_NO, K3D,   K3E,   KC_NO }, \
+    { K40,   K41,   k42, KC_NO,   KC_NO, KC_NO, KC_NO, K47,   KC_NO, KC_NO, K4A,   K4B,   KC_NO, K4D,   K4E,   K4F   }, \
+}
```
Comment (adophoxia) [inline keyboards/gray_studio/think65v3/think65v3.h:57]:
```suggestion

```
Matrix positions inside should be moved into `info.json` with this file being deleted after the process.

File: keyboards/gray_studio/think65v3/think65v3.c [hunk 4]
```diff
@@ -0,0 +1,3 @@
+// Copyright 2023 Yizhen Liu (@edwardslau)
+// SPDX-License-Identifier: GPL-2.0
+#include "think65v3.h"
```
Comment (adophoxia) [inline keyboards/gray_studio/think65v3/think65v3.c:3]:
```suggestion
#include "quantum.h"

const rgblight_segment_t PROGMEM my_capslock_layer[] = RGBLIGHT_LAYER_SEGMENTS(
    {0,2, HSV_WHITE}
);     
const rgblight_segment_t* const PROGMEM my_rgb_layers[] = RGBLIGHT_LAYERS_LIST(
    my_capslock_layer
);

void keyboard_post_init_kb(void) {
    // Enable the LED layers
    rgblight_layers = my_rgb_layers;
}
bool led_update_kb(led_t led_state) {
    if (!led_update_kb(led_state)) {return false;}
    rgblight_set_layer_state(0, led_state.caps_lock);
    return true;
}
```
* Should be renamed to `quantum.h` after the deletion of `think65v3.h`, assuming the matrix positions are moved to `info.json`.
* Move code from `keymap.c` to `think65v3.c`

File: keyboards/gray_studio/think65v3/think65v3.h [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    { k30, KC_NO, k32,   k33, k34,   k35,   k36,   k37, k38,   k39,   k3A, k3B, KC_NO, k3D,   k3E,   KC_NO }, \
+    { k40, k41,   k42, KC_NO, KC_NO, KC_NO, KC_NO, k47, KC_NO, KC_NO, k4A, k4B, KC_NO, k4D,   k4E,   k4F }, \
+}
+
+#define LAYOUT_65_iso_blocker( \
+    K00, K01, K02, K03, K04, K05, K06, K07, K08, K09, K0A, K0B, K0C,    K0D,   K0F,   \
+    K10,   K12, K13, K14, K15, K16, K17, K18, K19, K1A, K1B, K1C, K1D,         K1F,   \
+    K20,    K22, K23, K24, K25, K26, K27, K28, K29, K2A, K2B, K2C, K1E, K2D,          \
+    K30,  K31, K32, K33, K34, K35, K36, K37, K38, K39, K3A, K3B, K3D,     K3E,        \
+    K40,  K41,  K42,                    K47,           K4A,  K4B,     K4D, K4E, K4F    \
```
Comment (adophoxia) [inline keyboards/gray_studio/think65v3/think65v3.h:50]: Something to note of when porting to `info.json` as compiling the firmware from this PR as is right now gives out this warning:
```
QMK Firmware 0.20.1
☒ Invalid LAYOUT macro in keyboards/gray_studio/think65v3/think65v3.h: Key K42 in macro LAYOUT_65_iso_blocker has no matrix position!
Making gray_studio/think65v3 with keymap via

☒ Invalid LAYOUT macro in keyboards/gray_studio/think65v3/think65v3.h: Key K42 in macro LAYOUT_65_iso_blocker has no matrix position!
```

File: keyboards/gray_studio/think65v3/keymaps/default/keymap.c [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+);
+
+void keyboard_post_init_user(void) {
+    // Enable the LED layers
+    rgblight_layers = my_rgb_layers;
+}
+bool led_update_user(led_t led_state) {
+    rgblight_set_layer_state(0, led_state.caps_lock);
+    return true;
+}
```
Comment (adophoxia) [inline keyboards/gray_studio/think65v3/keymaps/default/keymap.c:49]:
```suggestion

```
If this is repeated in the `via` keymap, it'd be more practical to move this over to `think65v3.c` to function at a keyboard level.

File: keyboards/gray_studio/think65v3/keymaps/via/keymap.c [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+);
+
+void keyboard_post_init_user(void) {
+    // Enable the LED layers
+    rgblight_layers = my_rgb_layers;
+}
+bool led_update_user(led_t led_state) {
+    rgblight_set_layer_state(0, led_state.caps_lock);
+    return true;
+}
```
Comment (adophoxia) [inline keyboards/gray_studio/think65v3/keymaps/via/keymap.c:49]:
```suggestion

```
As above.

**Review by adophoxia — CHANGES_REQUESTED** [review]

**Comment by edwardslau** [comment]
@adophoxia Thank you for the review, I have fixed the format.

File: keyboards/gray_studio/think65v3/info.json [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+                {"label": "k42", "matrix": [4, 2],"x": 2.5, "y": 4, "w": 1.25},
+                {"label": "k47", "matrix": [4, 7],"x": 3.75, "y": 4, "w": 6.25},
+                {"label": "k4a", "matrix": [4, 10],"x": 10, "y": 4, "w": 1.25},
+                {"label": "k4b", "matrix": [4, 11],"x": 11.25, "y": 4, "w": 1.25},
+                {"label": "k4d", "matrix": [4, 13],"x": 13, "y": 4},
+                {"label": "k4e", "matrix": [4, 14],"x": 14, "y": 4},
+                {"label": "k4f", "matrix": [4, 15],"x": 15, "y": 4}
+            ]
+        },
+        "LAYOUT_65v3_ansi_blocker": {
```
Comment (waffle87) [inline keyboards/gray_studio/think65v3/info.json:118]:
```suggestion
        "LAYOUT_65_ansi_blocker": {
```

File: keyboards/gray_studio/think65v3/info.json [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+                {"label": "k42", "matrix": [4, 2],"x": 2.5, "y": 4, "w": 1.25},
+                {"label": "k47", "matrix": [4, 7],"x": 3.75, "y": 4, "w": 6.25},
+                {"label": "k4a", "matrix": [4, 10],"x": 10, "y": 4, "w": 1.25},
+                {"label": "k4b", "matrix": [4, 11],"x": 11.25, "y": 4, "w": 1.25},
+                {"label": "k4d", "matrix": [4, 13],"x": 13, "y": 4},
+                {"label": "k4e", "matrix": [4, 14],"x": 14, "y": 4},
+                {"label": "k4f", "matrix": [4, 15],"x": 15, "y": 4}
+            ]
+        },
+        "LAYOUT_65v3_iso_blocker": {
```
Comment (waffle87) [inline keyboards/gray_studio/think65v3/info.json:191]:
```suggestion
        "LAYOUT_65_iso_blocker": {
```

File: keyboards/gray_studio/think65v3/keymaps/default/keymap.c [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______, 
+        _______, _______, _______,                            _______,                   _______, _______,                   _______, _______, _______
+    ),
+    [3] = LAYOUT(
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
+        _______,          _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
+        _______,          _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______, 
+        _______, _______, _______,                            _______,                   _______, _______,                   _______, _______, _______
+    )
```
Comment (waffle87) [inline keyboards/gray_studio/think65v3/keymaps/default/keymap.c:33]:
```suggestion
    )
```

File: keyboards/gray_studio/think65v3/readme.md [hunk 11]
```diff
@@ -0,0 +1,2 @@
+# Gray Studio think65v3
+
```
Comment (waffle87) [inline keyboards/gray_studio/think65v3/readme.md:2]: This file needs to follow the [template](https://github.com/qmk/qmk_firmware/blob/master/data/templates/keyboard/readme.md).

File: keyboards/gray_studio/think65v3/think65v3.h [hunk 12]
```diff
@@ -0,0 +1,57 @@
+// Copyright 2023 Yizhen Liu (@edwardslau)
+// SPDX-License-Identifier: GPL-2.0
+#pragma once
+
+#include "quantum.h"
+
```
Comment (waffle87) [inline keyboards/gray_studio/think65v3/think65v3.h:6]: This file should be removed, as the `matrix` data is already in `info.json`.

**Review by waffle87 — CHANGES_REQUESTED** [review]

File: keyboards/gray_studio/think65v3/info.json (hunk 8 above)
Comment (edwardslau) [reply keyboards/gray_studio/think65v3/info.json:118]: This community layout LAYOUT_65_ansi_blocker is different from our layout(with a 2U blocker).

File: keyboards/gray_studio/think65v3/info.json (hunk 9 above)
Comment (edwardslau) [reply keyboards/gray_studio/think65v3/info.json:191]: As above

File: keyboards/gray_studio/think65v3/info.json (hunk 8 above)
Comment (waffle87) [reply keyboards/gray_studio/think65v3/info.json:118]: The community layout isn't actually being used. Either way, see the [warning that was pointed out above](https://github.com/qmk/qmk_firmware/pull/20280#discussion_r1151305124). Something like `LAYOUT_blocker` & `LAYOUT_iso_blocker` would be preferable.

Comment (edwardslau) [reply keyboards/gray_studio/think65v3/info.json:118]: > 

The community layout requires 67 keys, whereas our layout only has 65 keys. This does not compile.

Comment (waffle87) [reply keyboards/gray_studio/think65v3/info.json:118]: Then
> Something like `LAYOUT_blocker` & `LAYOUT_iso_blocker` would be preferable.

**Comment by edwardslau** [comment]
Will do some revisions and pr again

**Pull request closed without merging**